File: app/src/main/java/co/com/cesarnorena/pokedex/model/Pokemon.java
package co.com.cesarnorena.pokedex.model;

import com.google.gson.annotations.SerializedName;

/**
 * Created by Cesar Norena on 16/01/2016.
 *
 */
public class Pokemon {

    @SerializedName("id")
    private int id;

    @SerializedName("name")
    private String name;

    private String imageUrl;

    public Pokemon(int id, String name, String imageUrl) {
        this.id = id;
        this.name = name;
        this.imageUrl = imageUrl;
    }

    public int getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public String getImageUrl() {
        if (imageUrl == null)
            this.imageUrl = "http://assets.pokemon.com/assets/cms2/img/pokedex/full/"
                    + getFormattedId(id) + ".png";

        return imageUrl;
    }

    public static String getFormattedId(int id) {
        String formattedId;

        if (id < 10)
            formattedId = "00" + id;
        else if (id < 100)
            formattedId = "0" + id;
        else
            formattedId = String.valueOf(id);

        return formattedId;
    }
}

File: app/src/main/kotlin/com/cesarnorena/pokedex/data/model/Pokemon.kt
package com.cesarnorena.pokedex.data.model

import com.cesarnorena.pokedex.app.libraries.extensions.formattedId
import com.cesarnorena.pokedex.data.repository.remote.PokemonService
import com.google.gson.annotations.SerializedName

data class Pokemon(
    @SerializedName("id") val id: Int,
    @SerializedName("name") val name: String,
    @SerializedName("types") val typeSlots: List<TypeSlot>
) {
    val imageUrl: String get() = PokemonService.getLargeImageUrl(id.formattedId())
}

data class TypeSlot(
    @SerializedName("slot") val position: Int,
    @SerializedName("type") val type: Type
)

data class Type(
    @SerializedName("name") val name: String
)
